### PR TITLE
Adding atomicSet and atomicSetArray collection strategies

### LINKED ISSUE #1009: Doctrine take too much time to save/update a document
When document class contains a lot of properties, Doctrine save/update each property one by one this take too much time for response (in my case taken more than 2 minutes) throwing a time out error

**Document class definition**
```php
<?php

namespace MyNameSpace\Document;

use Doctrine\Bundle\MongoDBBundle\Validator\Constraints\Unique;
use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
use Gedmo\Mapping\Annotation as Gedmo;
use Gedmo\Timestampable\Traits\TimestampableDocument;
use JMS\Serializer\Annotation as Serializer;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * Profile
 *
 * @ODM\Document(
 *     indexes={
 *         @ODM\Index(keys={"user"="DES"}, options={"unique"=true})
 *     },
 *     repositoryClass="MyNameSpace\Repository\ProfileRepository"
 * )
 * @ODM\HasLifecycleCallbacks
 * @Unique(
 *     fields={"user"}
 * )
 *
 * @Serializer\XmlRoot("profile")
 */
class Profile {
    /**
     * @ODM\Id(strategy="none")
     * @Serializer\XmlAttribute
     * @Serializer\Groups({"simple"})
     */
    private $id;

    /**
     * @var \MyNameSpace\Document\User
     *
     * @ODM\ReferenceOne(targetDocument="User", simple=true)
     * @Assert\NotNull
     * @Assert\Type(type="object")
     * @Serializer\Groups({"simple"})
     */
    private $user;

    /**
     * @var string
     *
     * @ODM\String
     * @Serializer\Groups({"simple"})
     */
    private $companyName;

    /**
     * @var boolean
     *
     * @ODM\Boolean
     * @Serializer\Groups({"simple"})
     */
    private $isCompany = false;

    /**
     * @var string
     *
     * @ODM\String
     * @Serializer\Groups({"simple"})
     */
    private $title;

    /**
     * @var \MyNameSpace\Document\Country
     *
     * @ODM\ReferenceOne(targetDocument="Country", simple=true)
     * @Assert\NotNull
     */
    private $country;

    /**
     * @var \MyNameSpace\Document\City
     *
     * @ODM\ReferenceOne(targetDocument="City", simple=true)
     * @Assert\NotNull
     */
    private $city;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Length(min=0, max=1000)
     */
    private $aboutMe;

    /**
     * @var \MyNameSpace\Document\Interest
     *
     * @ODM\EmbedOne(targetDocument="Interest")
     */
    private $interest;

    /**
     * @var \MyNameSpace\Document\Tag
     *
     * @ODM\EmbedMany(targetDocument="Tag")
     * @Assert\Count(min="0", max="10")
     */
    private $keywords;

    /**
     * @var boolean
     *
     * @ODM\Boolean
     */
    private $located = false;

    /**
     * @var \MyNameSpace\Document\Coordinate
     *
     * @ODM\EmbedOne(targetDocument="Coordinate")
     */
    private $hotspot;

    /**
     * @var float
     *
     * @ODM\Distance
     */
    private $distance;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $website;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $vimeo;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $twitter;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $imdb;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $facebook;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $youtube;

    /**
     * @var string
     *
     * @ODM\String
     * @Assert\Url
     */
    private $linkedin;

    /**
     * @var \MyNameSpace\Document\Cover
     *
     * @ODM\EmbedMany(targetDocument="Cover")
     * @Assert\Count(min="0", max="3")
     */
    private $covers;

    /**
     * @var \MyNameSpace\Document\Recomendation
     *
     * @ODM\EmbedMany(targetDocument="Recomendation")
     */
    private $recomendations;

    /**
     * @var \MyNameSpace\Document\Project
     *
     * @ODM\ReferenceMany(targetDocument="Project", simple=true)s
     */
    private $projects;

    /**
     * @var \MyNameSpace\Document\Profile
     *
     * @ODM\ReferenceMany(targetDocument="Profile", simple=true)
     */
    private $following;

    /**
     * @var \MyNameSpace\Document\Profile
     *
     * @ODM\ReferenceMany(targetDocument="Profile", simple=true)
     */
    private $followers;

    /**
     * @var \MyNameSpace\Document\Project
     *
     * @ODM\ReferenceMany(targetDocument="Project", simple=true)
     */
    private $watching;

    /**
     * Hook timestampable behavior
     * updates createdAt, updatedAt fields
     */
    use TimestampableDocument;


    public function __construct()
    {
        $this->keywords       = new \Doctrine\Common\Collections\ArrayCollection();
        $this->covers         = new \Doctrine\Common\Collections\ArrayCollection();
        $this->recomendations = new \Doctrine\Common\Collections\ArrayCollection();
        $this->projects       = new \Doctrine\Common\Collections\ArrayCollection();
        $this->following      = new \Doctrine\Common\Collections\ArrayCollection();
        $this->followers      = new \Doctrine\Common\Collections\ArrayCollection();
        $this->watching       = new \Doctrine\Common\Collections\ArrayCollection();
    }

   // Setters & Getters

    /**
     * @ODM\PreFlush
     */
    public function prepare() {
        if(!$this->located) {
            $this->hotspot = null;
        }
        if(count($this->keywords) < 1) {
            $this->keywords = null;
        }
        if(count($this->covers) < 1) {
            $this->covers = null;
        }
        if(count($this->recomendations) < 1) {
            $this->recomendations = null;
        }
        if(count($this->projects) < 1) {
            $this->projects = null;
        }
        if(count($this->following) < 1) {
            $this->following = null;
        }
        if(count($this->followers) < 1) {
            $this->followers = null;
        }
        if(count($this->watching) < 1) {
            $this->watching = null;
        }
    }
}
```

**Documet Manager**
```php
<?php

namespace MyNameSpace\Model;

/**
 * DocumentManager
 */
class DocumentManager {

    /**
     * MongoDB Document Manager
     *
     * @var \Doctrine\ODM\MongoDB\DocumentManager
     */
    private $dm;

    /**
     * MongoDB Document Repository
     *
     * @var \Doctrine\ODM\MongoDB\DocumentRepository
     */
    private $repository;

    /**
     *
     * @param \Doctrine\ODM\MongoDB\DocumentManager $dm Doctrine  MongoDB instance
     * @param string                                $documentName name of document
     */
    public function __construct(\Doctrine\ODM\MongoDB\DocumentManager $dm, $documentName) {
        $this->dm = $dm;
        if ($dm) {
            $this->repository = $dm->getRepository($documentName);
        }
    }

    /**
     * Fetch all document in a collection
     *
     * @param int $limit How many documents to return.
     * @param int $offset Offset from which to start listing documents.
     *
     * @return \Doctrine\MongoDB\Collection
     */
    public function fetch($limit = null, $offset = null) {
        return $this->repository->findBy(array(), null, $limit, $offset);
    }

    /**
     * Find one document in a collection
     *
     * @param string $id Document identifier
     *
     * @return object MongoDB document
     */
    public function get($id) {
        return $this->repository->findOneById($id);
    }

    /**
     * Find documents according a criteria
     *
     * @param  array  $criteria
     * @return \Doctrine\MongoDB\Collection
     */
    public function find($criteria = array(), $limit = null, $offset = null)
    {
        return $this->repository->findBy($criteria, null, $limit, $offset);
    }

    /**
     * Save document to MongoDB
     *
     * @param object $document MongoDB document
     */
    public function save($document) {
        $this->dm->persist($document);
        $this->dm->flush();
    }

    /**
     * Update document to MongoDB
     *
     * @param object $document MongoDB document
     */
    public function update() {
        $this->dm->flush();
    }

    /**
     * Delete document from MongoDB
     *
     * @param  object $document MongoDB document
     * @return boolean true when succes
     */
    public function remove($document) {
        $this->dm->remove($document);
        $this->dm->flush();

        return true;
    }

    /**
     * Get Document Repository
     *
     * @return \Doctrine\ODM\MongoDB\DocumentRepository
     */
    public function getRepository() {
        return $this->repository;
    }

    /**
     * Find one document in a collection
     *
     * @param  \Doctrine\ODM\MongoDB\PersistentCollection $collection
     * @param  string                                     $id         Document id
     * @return object                                     $document   MongoDB document
     */
    public function findOneInCollection(\Doctrine\ODM\MongoDB\PersistentCollection $collection, $id)
    {
        $document = $collection->filter(
            function($item) use ($id) {
                return $item->getId() == $id;
            }
        );
        return $document->first();
    }
}
```

**Save/update action on controller**
```php
<?php

namespace MyNameSpace\Controller;

use FOS\RestBundle\Controller\Annotations as REST;
use FOS\RestBundle\Request\ParamFetcherInterface;
use FOS\RestBundle\Util\Codes;
use Nelmio\ApiDocBundle\Annotation\ApiDoc;
use MyNameSpace\Document\Profile;
use MyNameSpace\Form\ProfileType;
use Symfony\Component\HttpFoundation\Request;

/**
 * Profile API
 *
 * @REST\RouteResource("Profile")
 */
class ProfileController extends BaseController
{
// Some actions

    /**
     * Creates a new profile from the submitted data.
     *
     * @param Request $request the request object
     *
     * @return \Symfony\Component\Form\FormTypeInterface|\FOS\RestBundle\View\View
     *
     * @ApiDoc(
     *     resource    = true,
     *     description = "Create profile",
     *     input       = "MyNameSpace\Form\ProfileType",
     *     output      = "MyNameSpace\Document\Profile",
     *     statusCodes = {
     *         201 = "Returned when a new profile is created",
     *         400 = "Returned when error data was sended",
     *         403 = "Returned when is not authorized"
     *     }
     * )
     *
     * @REST\View(
     *     templateVar = "form",
     *     statusCode  = Codes::HTTP_BAD_REQUEST
     * )
     */
    public function postAction(Request $request) {
        $document = new Profile();
        $form     = $this->createForm(new ProfileType(), $document);

        $form->submit($request);
        if ($form->isValid()) {
            $this->get('rest_api.dm.profile')->save($document);

            $view = $this->view($document, 201)
                // set the `Location` header only when creating new resources
                ->setLocation($this->generateUrl(
                    'nu_get_profile', array('id' => $document->getId()),
                    true // absolute
                ));

            return $view;
        }

        return $form;
    }

    /**
     * Update existing profile or create new from the submitted data.
     *
     * @param Request $request the request object
     * @param int     $id      the profile id
     *
     * @return \Symfony\Component\Form\FormTypeInterface|\FOS\RestBundle\View\View
     *
     * @ApiDoc(
     *     resource     = true,
     *     description  = "Update profile",
     *     input        = "MyNameSpace\Form\ProfileType",
     *     output       = "MyNameSpace\Document\Profile",
     *     statusCodes  = {
     *         201 = "Returned when a new resource is created",
     *         204 = "Returned when successful",
     *         400 = "Returned when the data has errors",
     *         403 = "Returned when is not authorized"
     *     },
     *     requirements = {
     *         {
     *             "name"        = "id",
     *             "dataType"    = "string",
     *             "requirement" = "[a-f0-9]{24}",
     *             "description" = "Profile identifier"
     *         }
     *    }
     * )
     *
     * @REST\View(
     *     templateVar = "form",
     *     statusCode  = Codes::HTTP_BAD_REQUEST
     * )
     */
    public function putAction(Request $request, $id) {
        $document   = $this->get('rest_api.dm.profile')->get($id);
        $statusCode = Codes::HTTP_NO_CONTENT;

        if (!$document) {
            $document   = new Profile();
            $statusCode = Codes::HTTP_CREATED;
        }

        $form = $this->createForm(new ProfileType(), $document);

        $form->submit($request);
        if ($form->isValid()) {
            if($statusCode === Codes::HTTP_CREATED)
                $this->get('rest_api.dm.profile')->save($document);
            else
                $this->get('rest_api.dm.profile')->update();

            $view = $this->view($document, $statusCode)
                ->setRoute('nu_get_profile')
                ->setRouteParameters(array('id' => $document->getId()));

            return $view;
        }

        return $form;
    }
// Other actions
}
```

Is possible to save/update all fields at once? and how can I configure this?

### PR DESCRIPTION
**Pull request by malarzm**
This is an early implementation of how ``atomicSet`` and ``atomicSetArray`` could work in my opinion. There are some things that needs to be done

- [x] ~~maybe add `atomic=true` flag to mapping instead of new strategy (or change existing `$set` behaviour)~~
- [x] inserts
- [x] updates
- [x] upserts
- [x] removing collection
- [x] ~~collection within collection with different strategies~~ throw exception if used in embedded documents
- [x] more tests
- [x] docs

cc @jwage @jmikola @alcaeus @lavoiesl @blockjon

If this PR will reach its final form it will resolve #1009, #437, #1095 and #1094

### CODE REVIEW

**Comment by blockjon** [comment]
@malarzm I was thinking that the ODM should *always* perform all changes to a record in one atomic query. Do you agree, or do you do you feel as though this behavior should be parameterized?

**Comment by malarzm** [comment]
ODM can't do all changes within one query, for instance (citing @jmikola from https://github.com/doctrine/mongodb-odm/issues/1009#issuecomment-78718247)

> Collections are currently handled separately out of necessity for the base case where elements are being added and removed from the same array, since MongoDB doesn't allow that in the same write operation. 

but of course this in not the case with `$set`. For now I have made this parametrized as I wanted to have current tests pass and only new ones I'm writing fail but changing this will not be hard if we come to conclusion that we want to change existing behaviour.

**Comment by malarzm** [comment]
I'm afraid we won't be able to change existing `$set` behaviour since I'm not able to make atomic update from within ``DocumentPersister`` to collection nested in another collection with `pushAll` strategy (especially not when embedded object containing that collection is yet to be created). Right now I'm thinking about limiting the feature to be available only in top level document (or deeply embedded but only within collections with `atomic*` strategy).

I've been also thinking that `atomic` flag in `set` and `setArray` could adapt and make atomic updates when possible and fall back to old behaviour but I think this can be at least confusing. That said I think that introducing new strategies is ok and even wanted given said limitations.

**Comment by malarzm** [comment]
@blockjon could you run your test against this PR?

**Comment by blockjon** [comment]
Having just quickly read through the code, it appears as if this might do the trick. I will try this today and report back my findings.

**Comment by blockjon** [comment]
@malarzm Looks like your changes fix my problem as seen here: https://github.com/malarzm/mongodb-odm/pull/3

File: docs/en/reference/collection-strategies.rst [hunk 1]
```diff
@@ -45,13 +45,34 @@ strategy, but will first numerically reindex the collection to ensure that it is
...
+
+The ``atomicSet`` strategy uses MongoDB's `$set`_ operator to update the entire
+collection with a single update query. Unlike with ``set`` strategy there will
+be only one query for updating both parent document and collection itself. This
+strategy can be especially useful when dealing with high concurrency and 
+:ref:`versioned documents <annotations_reference_version>`.
+
+.. note::
+
+    Only top level document can contain collection with ``atomicSet`` or
```
Comment (jmikola) [inline docs/en/reference/collection-strategies.rst]: I'd rephrase as "The `atomicSet` and `atomicSetArray` strategies may only be used for collections on a top-level document."

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/AtomicSetTest.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    public $surname;
+    
+    /** @ODM\EmbedMany(strategy="atomicSet",targetDocument="Documents\Phonenumber") */
+    public $phonenumbers;
+    
+    public function __construct($name)
+    {
+        $this->name = $name;
+        $this->phonenumbers = new ArrayCollection();
+    }
```
Comment (jmikola) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/AtomicSetTest.php]: I think we can add an additional test here with a nested embed-many relationship. Perhaps: AtomicUser embeds many PhoneBooks, which each embed many Phonenumbers. The top-level PhoneBooks relationship will use atomicSet, and the nested Phonenumbers relationship can use the default strategy.

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 3]
```diff
@@ -235,6 +228,11 @@ public function executeInserts(array $options = array())
                 }
                 $data[$versionMapping['name']] = $nextVersion;
             }
+            
+            $atomicCollectionQuery = $this->getAtomicCollectionUpdateQuery($document);
+            if (!empty($atomicCollectionQuery['$set'])) {
+                $data = array_merge_recursive($data, $atomicCollectionQuery['$set']);
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Why a recursive merge here? I'd like to think that the collection's field name wouldn't conflict with any other mapped fields on the top-level document, so the array keys won't conflict. And if they did conflict (which I'd consider a logical mapping error), a recursive merge might lead to some odd behavior where we're merging data within the collection from two different sources.

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 4]
```diff
@@ -282,6 +280,8 @@ public function executeUpserts(array $options = array())
                 }
                 $data['$set'][$versionMapping['name']] = $nextVersion;
             }
+            
+            $data = array_merge_recursive($data, $this->getAtomicCollectionUpdateQuery($document));
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: I suppose the recursive merge here is because `getAtomicCollectionUpdateQuery()` may return `$set` or `$unset` operations, and it keeps the code simpler than doing a non-recursive merge on `$data['$set']` or `$data['$unset']` (and having to ensure that `$data` has those array keys, too)?

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 3 above)
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Per my [comment below](https://github.com/doctrine/mongodb-odm/pull/1096/files#r30641740) on `executeUpserts()`, it occurred to me that `getAtomicCollectionUpdateQuery()` might populate `$unset`. Should we have a sanity check on that and raise an exception if so, or is it not worth the trouble?

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 4 above)
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: I would suggest optimizing this to avoid the unnecessary function call, similar to what you did for `executeInserts()`:

```
$atomicCollectionQuery = $this->getAtomicCollectionUpdateQuery($document);
if ( ! empty($atomicCollectionQuery) {
    $data = array_merge_recursive($data, $atomicCollectionQuery);
}
```

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 5]
```diff
@@ -375,6 +375,8 @@ public function update($document, array $options = array())
                     $this->class->reflFields[$this->class->versionField]->setValue($document, $nextVersion);
                 }
             }
+            
+            $update = array_merge_recursive($update, $this->getAtomicCollectionUpdateQuery($document));
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Same suggested optimization as https://github.com/doctrine/mongodb-odm/pull/1096/files#r30642646.

**Comment by blockjon** [comment]
Friendly question for you all - This branch fixes a critical problem I'm faced with and I would love to use it sooner rather than later.

Since I'm new to this project, I'd like to get a sense of when you feel as though it would be in a state that you feel I should attempt to integrate into my project. Should I want until this gets merged to master, or until you tag a beta 13, for example?

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 6]
```diff
@@ -1230,4 +1232,31 @@ private function getClassDiscriminatorValues(ClassMetadata $metadata)
...
+            $collPersister = $this->uow->getCollectionPersister();
+            if ($this->uow->isCollectionScheduledForUpdate($coll)) {
+                $update = array_merge_recursive($update, $collPersister->prepareSetQuery($coll));
+                $this->uow->unscheduleCollectionUpdate($coll);
+                /* Collection can be set for both deletion and update if
+                 * PersistentCollection instance was changed. Since we're dealing
+                 * with collection update in one query we won't need the $unset
+                 */
+                $this->uow->unscheduleCollectionDeletion($coll);
+            } else {
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Would it be safer swap out the `else` for a separate `isCollectionScheduledForDeletion()` check here? I realize `UnitOfWork::scheduleCollectionOwner()`, which populates the array returned by `getScheduledCollections()`, is currently only called from `scheduleCollectionDeletion()` and `scheduleCollectionUpdate()`; however, if that changes in the future, we're looking at a very destructive `else` condition :)

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 7]
```diff
@@ -1230,4 +1232,31 @@ private function getClassDiscriminatorValues(ClassMetadata $metadata)
...
+            }
+            $collPersister = $this->uow->getCollectionPersister();
+            if ($this->uow->isCollectionScheduledForUpdate($coll)) {
+                $update = array_merge_recursive($update, $collPersister->prepareSetQuery($coll));
+                $this->uow->unscheduleCollectionUpdate($coll);
+                /* Collection can be set for both deletion and update if
+                 * PersistentCollection instance was changed. Since we're dealing
+                 * with collection update in one query we won't need the $unset
+                 */
+                $this->uow->unscheduleCollectionDeletion($coll);
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Is this a redundant safety check? Would it be safer would a collection be scheduled for both update and deletion?

Why would a collection be scheduled for both update and deletion? If that a logic error we should instead check in UnitOfWork? `scheduleCollectionDeletion()` and `scheduleCollectionUpdate()` don't seem to care, but perhaps they should throw if the collection is already scheduled, or one should take precedence. For instance, if something schedules a collection for deletion, perhaps any update schedule should be ignored.

File: lib/Doctrine/ODM/MongoDB/UnitOfWork.php [hunk 8]
```diff
@@ -2752,8 +2799,12 @@ private function scheduleCollectionOwner(PersistentCollection $coll)
             list(, $document, ) = $this->getParentAssociation($document);
             $class = $this->dm->getClassMetadata(get_class($document));
         }
+        $oid = spl_object_hash($document);
 
-        $this->hasScheduledCollections[spl_object_hash($document)] = true;
+        if (!isset($this->hasScheduledCollections[$oid])) {
+            $this->hasScheduledCollections[$oid] = array();
+        }
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/UnitOfWork.php]: Was this conditional required? if `$this->hasScheduledCollections` is only ever null or an array, there should be no problem using the `[]` append operator. See: http://3v4l.org/5VikE

**Comment by malarzm** [comment]
@blockjon idea is to have this feature merged and tag BETA-13 then

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 7 above)
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: @malarzm mentioned this was due to #1011, which appears to be a separate issue. Perhaps we should add a TODO here to remove this once that issue is resolved, then. Long-term, I don't think ODM should be issuing superfluous `$unset` queries.

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 6 above)
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: The `elseif` was there before but I'll bring it back ;)

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 4 above)
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: For the record that's exactly reason for doing recursive merge :+1: Will add optimizations

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 6 above)
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: @jwage: If we make this an `elseif` again, how do you feel about a final `else` condition throwing a LogicException?

**Comment by jmikola** [comment]
LGTM pending test results. @jwage: Do you have a chance to take a look at this within the next day?

Ideally, we'd like to get it included in a new 1.0.0-BETA13 tag this week (to help with doctrine/DoctrineMongoDBBundle#298).

**Comment by jwage** [comment]
:+1:

**Comment by blockjon** [comment]
I AM EXCITED!!!!!!

**Comment by smolinari** [comment]
Nice work gentlemen. :+1: 

Scott

**Comment by BernardoSilva** [comment]
:+1: Nice work.

**Comment by malarzm** [comment]
@blockjon BETA13 is now released :)

**Comment by blockjon** [comment]
@malarzm This is fantastic news thank you!!!!

**Comment by daniphp** [comment]
@malarzm - I think you need to implement atomicSet as default where possible. I just found a bug in production that happened in only 1/100.000 cases and it was driving me nuts for weeks before I found the problem. Using pushAll as default is really something unexpected for me as a developer.

**Comment by blockjon** [comment]
@daniphp +1 on your comments there... I had the same feedback here https://github.com/doctrine/mongodb-odm/pull/782#issuecomment-113642286
